docs/code-culture.md

```diff
@@ -164,3 +164,66 @@ Example of good framework log messages (a little bit simplified) can look like e
+
+### On Libraries and Styles
+
+We heavily utilize [ScalaTest](http://www.scalatest.org/) as our primary test library. When it comes to the test style we prefer [WordSpec](http://www.scalatest.org/at_a_glance/WordSpec)
+
+```scala
+  // Describe a scope for a subject, in this case: "A Set"
+  "A Set" can { // All tests within these curly braces are about "A Set"
+
+    // Can describe nested scopes that "narrow" its outer scopes
+    "empty" should { // All tests within these curly braces are about "A Set (when empty)"
+
+      "have size 0" in {    // Here, 'it' refers to "A Set (when empty)". The full name
+        Set.empty.size shouldBe 0 // of this test is: "A Set (when empty) should have size 0"
+      }
+      ...
+```
+
+combined with [Given, When, And Then](http://www.scalatest.org/getting_started_with_feature_spec) and scala [matchers](http://www.scalatest.org/user_guide/using_matchers):
+
+```scala
+  "A mutable Set" should "allow an element to be added" in {
+    Given("an empty mutable Set")
+    val set = mutable.Set.empty[String]
+
+    When("an element is added")
+    set += "clarity"
+
+    Then("the Set should have size 1")
+    set.size shouldBe 1
+
+    And("the Set should contain the added element")
+    set should contain theSameElementsAs Set("clarity")
+  }
+```
+
+which give us an opportunity to write a test as a short story documenting part of the system behaviour.
+
+### On Testing Granularity
+
+In USI we define following granularity levels:
+
+* **Unit tests**: the lowest test level possible. Unit test are cheap, can be run fast and give the developer an immediate feedback on when something is obviously broken
+* **Integration tests**: an integration test starts a minimal Mesos cluster consisting of Mesos master, Mesos agent, in-memory Zookeeper server and a framework. These are more expensive tests that typically require more resources and time to run. Tests that require back and forth communication with Mesos are best placed here because we don't want to mock Mesos responses.
+* **System integraiton test**: here we start a full-blown DC/OS cluster testing all aspects of framework interacting with the DC/OS ecosystem. These are the most expensive tests (in terms of time and money) that would typically reqire a DC/OS cluster running on e.g. AWS utilizing multiple EC2 nodes, volumes, ELB etc. A system integration test would typically cover some coarse-grained USI feature or a feature that relies on other DC/OS components like the secret store.
+
+**As a rule of thumb a broken behavior should fail at the lowest possible level**. Most features will be covered on more than one level but the coverage is different. Let's consider support for [Mesos fetcher](http://mesos.apache.org/documentation/latest/fetcher/) as an example. The Mesos fetcher is a mechanism to download resources into the sandbox directory of a task in preparation of running the task. So how do the tests look like?
```

Comment:
```suggestion
**As a rule of thumb, a broken behavior should fail at the lowest possible level**. Most features will be covered on more than one level but the coverage is different. Let's consider support for [Mesos fetcher](http://mesos.apache.org/documentation/latest/fetcher/) as an example. In preparation to run a task, the Mesos fetcher downloads resources into the task's sandbox directory. So how do the tests look like?
```

Reply:
Funny, this sentence about Mesos fetcher is taken directly from [Mesos documentation](http://mesos.apache.org/documentation/latest/fetcher/#what-is-the-mesos-fetcher)